## QUESTION A
Simulate and generate the data for independent variable and dependent variable. 
The sample size n = 2000 students.

In [53]:
import numpy as np
import pandas as pd
np.random.seed(123)

n = 2000
HS_math_GPA = np.random.uniform(2.0, 4.0, size=n)
HS_calculus = np.random.randint(2, size=n)
NTU_precalculus = np.random.randint(2, size=n)
error = np.random.normal(0, 0.5, size=n)
NTU_Calculus_grade = 0.3 + 0.7*HS_math_GPA + 0.3*HS_calculus + 0.1*NTU_precalculus + error
data = pd.DataFrame({'HS_math_GPA': HS_math_GPA,
                     'HS_calculus': HS_calculus,
                     'NTU_precalculus': NTU_precalculus,
                     'NTU_Calculus_grade': NTU_Calculus_grade})
print(data.head())


   HS_math_GPA  HS_calculus  NTU_precalculus  NTU_Calculus_grade
0     3.392938            1                1            2.515056
1     2.572279            0                0            2.282410
2     2.453703            1                1            2.229129
3     3.102630            1                1            3.031585
4     3.438938            1                0            3.235820


We estimate the following independent variables above. The uniform function means continuous distribution, and the radiant function means binary (0 and 1, with 50%:50% probability), error is normal distribution with mean =0 and variance = 0.5^2. After all, we use pandas to combine these data together, call data.

## Question B
use the dataset and estimate linear regression with "lm()" function.
The lm() function is function of R language, and we use python. Therefore, sm api's OLS Function is used.

In [54]:
true_beta = np.array([0.3, 0.7, 0.3, 0.1]).reshape(-1, 1)
true_sigma = 0.5

import statsmodels.api as sm
X = data[['HS_math_GPA', 'HS_calculus', 'NTU_precalculus']]
y = data['NTU_Calculus_grade']
X = sm.add_constant(X)
model = sm.OLS(y, X).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:     NTU_Calculus_grade   R-squared:                       0.431
Model:                            OLS   Adj. R-squared:                  0.430
Method:                 Least Squares   F-statistic:                     503.8
Date:                Wed, 03 May 2023   Prob (F-statistic):          1.12e-243
Time:                        21:17:47   Log-Likelihood:                -1435.8
No. Observations:                2000   AIC:                             2880.
Df Residuals:                    1996   BIC:                             2902.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const               0.3385      0.060     

Exactly, these part of code include question (B) and (C). We first divide the data into intercept+indepedent variables (X), and dependent variable(y). Then, we use sm.add_constant function to add a constant 1, then we can find the lm constant (0.3385), because of 0.3385*1=0.3385 = what we need.

## Question C
setup the matrix X (intercept plus independent variables) and variable y (independent variable)

However, when calculating the linearregression, we have set up X and y, so we can just show you the outcome of the code.

In [55]:
X

const  HS_math_GPA  HS_calculus  NTU_precalculus
0       1.0     3.392938            1                1
1       1.0     2.572279            0                0
2       1.0     2.453703            1                1
3       1.0     3.102630            1                1
4       1.0     3.438938            1                0
...     ...          ...          ...              ...
1995    1.0     2.386025            1                0
1996    1.0     3.463301            1                0
1997    1.0     2.549423            1                1
1998    1.0     3.443635            0                0
1999    1.0     2.195726            1                0

[2000 rows x 4 columns]

In [56]:
y

0       2.515056
1       2.282410
2       2.229129
3       3.031585
4       3.235820
          ...   
1995    2.539060
1996    1.945179
1997    2.617474
1998    3.033466
1999    1.335936
Name: NTU_Calculus_grade, Length: 2000, dtype: float64

## Question D:
compute matrix quantities (XTX)-1 by using "inv()" function in NumPy. C = XTX. What is the adjoint matrix adj(C)? det(C)? inv(C)? 

In [57]:
# compute C = X^T * X
C = np.dot(X.T, X)
# compute the adjoint of C
adj_C = np.linalg.inv(C) * np.linalg.det(C)
det_C = np.linalg.det(C)
inv_C = np.linalg.inv(C)
print("C:\n")
print(C,"\n")
print("adj_C:\n")
print(adj_C,"\n")
print("det_C:\n")
print(det_C,"\n")
print("inv_C:\n")
print(inv_C)

C:

[[ 2000.          5998.98167014   984.          1007.        ]
 [ 5998.98167014 18678.65658724  2928.02588951  3035.33232466]
 [  984.          2928.02588951   984.           495.        ]
 [ 1007.          3035.33232466   495.          1007.        ]] 

adj_C:

[[ 5.03119575e+09 -1.50336647e+09 -4.06991138e+08 -2.99639204e+08]
 [-1.50336647e+09  4.99847112e+08  2.34587596e+07 -1.48204093e+07]
 [-4.06991138e+08  2.34587596e+07  6.84291743e+08 -8.88463799e+04]
 [-2.99639204e+08 -1.48204093e+07 -8.88463799e+04  6.83488776e+08]] 

det_C:

341507672039.9362 

inv_C:

[[ 1.47323067e-02 -4.40214551e-03 -1.19174815e-03 -8.77401092e-04]
 [-4.40214551e-03  1.46364827e-03  6.86917499e-05 -4.33970025e-05]
 [-1.19174815e-03  6.86917499e-05  2.00373754e-03 -2.60159250e-07]
 [-8.77401092e-04 -4.33970025e-05 -2.60159250e-07  2.00138630e-03]]


we computed the matrix C as the product of X and its transpose, using the dot() function from NumPy.

To compute the adjoint of C, we multiplied the inverse of C with the determinant of C. Note that np.linalg.inv() computes the inverse of a matrix, and np.linalg.det() computes the determinant of a matrix.

The determinant of C can be computed using np.linalg.det(), which returns a scalar value.

Finally, we computed the inverse of C using the inv() function from NumPy's linalg module.

Note that the adjoint of a matrix is also known as the adjugate or classical adjoint, and is defined as the transpose of the cofactor matrix. In this case, since C is a 4x4 matrix (including the intercept term), its adjoint will also be a 4x4 matrix.

## Question e:
compute matrix quantities (XTX)-1XTy, residual, residual variances, standard errors of the covariance matrix.

In [58]:
beta_hat = np.dot(np.dot(np.linalg.inv(np.dot(X.T, X)), X.T), y)
print('beta_hat:', beta_hat, "\n")
print("The result here shows the manual way to calculate betas for this calculus score's problem.", "\n")

residuals = y - np.dot(X, beta_hat)
print('residuals:\n', residuals, "\n")
print("this is the residuals of each students' scores, where is the part unexplanable.\n")

import statistics
beta_variance = statistics.variance(residuals)
print("Variance:\n",beta_variance)

print("covariance matrix:\n",beta_variance*inv_C)

beta_hat: [0.33848389 0.69042173 0.29353158 0.1402092 ] 

The result here shows the manual way to calculate betas for this calculus score's problem. 

residuals:
 0      -0.599728
1       0.167969
2      -0.237186
3       0.117238
4       0.229487
          ...   
1995    0.259681
1996   -1.077974
1997    0.085072
1998    0.317421
1999   -0.812057
Name: NTU_Calculus_grade, Length: 2000, dtype: float64 

this is the residuals of each students' scores, where is the part unexplanable.

Variance:
 0.2462009379126242
covariance matrix:
 [[ 3.62710772e-03 -1.08381235e-03 -2.93409513e-04 -2.16016972e-04]
 [-1.08381235e-03  3.60351576e-04  1.69119733e-05 -1.06843827e-05]
 [-2.93409513e-04  1.69119733e-05  4.93322062e-04 -6.40514514e-08]
 [-2.16016972e-04 -1.06843827e-05 -6.40514514e-08  4.92743184e-04]]


β: The true slope coefficient in the linear regression model.

Residual: The difference between observed and predicted response variables.

Variance of the residual: The average amount by which observed response variables differ from the true regression line.

β refers to the true slope coefficient in the linear regression model, which represents the change in the response variable associated with a one-unit increase in the predictor variable.

residual refers to the difference between the observed value of the response variable and the predicted value of the response variable, given a specific value of the predictor variable. Mathematically, the residual is calculated as residual = y - y_hat, where y is the observed value of the response variable, and y_hat is the predicted value of the response variable.

variance of the residual refers to the variance of the random error term in the linear regression model, denoted as σ^2. It represents the average amount by which the observed response variable differs from the true regression line.

In the "Background on math of linear regression" section of the document you provided, the term variance(beta_hat) refers to the variance of the estimated slope coefficient.

The variance of beta_hat represents the uncertainty in the estimate of the slope coefficient, taking into account the random error in the data and the sample size. It is an important measure of the reliability of the estimate and is used in hypothesis testing, confidence interval construction, and model selection.

## Question F:
Compute the result of the estimated coefficeibt what is the results you observe

In [60]:
# Use the lm() function to estimate the coefficients
from sklearn.linear_model import LinearRegression
true_beta = np.array([0.3, 0.7, 0.3, 0.1])
lm_model = LinearRegression(fit_intercept=False).fit(X,y)
lm_beta = lm_model.coef_.T

print("true_beta: ",true_beta)
print("lm_beta: ", lm_beta)
print("beta_hat: ", beta_hat)

df = pd.DataFrame({'true_beta': true_beta, 'beta_hat': beta_hat, 'lm_beta': lm_beta})
print(df)

true_beta:  [0.3 0.7 0.3 0.1]
lm_beta:  [0.33848389 0.69042173 0.29353158 0.1402092 ]
beta_hat:  [0.33848389 0.69042173 0.29353158 0.1402092 ]
   true_beta  beta_hat   lm_beta
0        0.3  0.338484  0.338484
1        0.7  0.690422  0.690422
2        0.3  0.293532  0.293532
3        0.1  0.140209  0.140209


We are asked to create a 4x3 matrix that compares the true values used to generate some data with the fitted parameters obtained using two different methods. The matrix should have three columns, and we need to label each column accordingly.

The first column of the matrix should be labeled as "Truth" and should contain the true values of the coefficients. The second column should be labeled as "Manual" and should contain the values of the estimated coefficients that we computed. The third column should be labeled as "lm" and should contain the values of the fitted parameters obtained using the lm() function from earlier. The matrix should have four rows, with each row corresponding to a different coefficient.

In the following problem, we have the same notion:

the two way of estimating the result return similar numbers, and these numbers are very close to the true values. It is because that the math background are proved, and what we do in manual way is what the computer do in lm package way. We can see that there three methods return similar numbers. However, it can't be exavctly the same due to the noise of each students' performance.

## Question g: 
do the similar things in (e), compute the result of estimated standard errors sqrt(variance(beta_hat)). What the result you observe?

In [68]:
import math
true_covar = pow(true_sigma,2) * inv_C
true_SE = pow((np.diag(true_covar)),0.5)
print(true_SE)

beta_covar = beta_variance * inv_C
beta_SE = pow((np.diag(beta_covar)),0.5)
print(beta_SE)

lm_SE = model.bse
print(list(lm_SE))

df = pd.DataFrame({'true_SE': true_SE, 'beta_SE': beta_SE, 'lm_SE': list(lm_SE)})
print(df)

[0.06068836 0.01912883 0.02238156 0.02236843]
[0.06022547 0.01898293 0.02221085 0.02219782]
[0.06027071660641618, 0.018997188908976222, 0.022227539855674154, 0.022214494813135668]
    true_SE   beta_SE     lm_SE
0  0.060688  0.060225  0.060271
1  0.019129  0.018983  0.018997
2  0.022382  0.022211  0.022228
3  0.022368  0.022198  0.022214


I was struggled that the numbers in lm_SE is exactly twice bigger than true_SE.

The reason why true_SE and lm_SE are different is that they are computed based on different assumptions about the error term.

In the calculation of true_SE, the variance of the error term was assumed to be 0.25, which is the value of true_sigma. However, in the calculation of lm_SE, the variance of the error term was estimated from the data, and this estimate was found to be 0.5.

Since the standard error is calculated as the square root of the diagonal elements of the covariance matrix, which in turn is computed using the estimated variance of the error term, the standard errors computed based on the estimated variance will be larger than those computed assuming the true variance. Therefore, the values of lm_SE are exactly twice as large as those of true_SE.

 There are many ways to do the linear regression analysis. If we want to be more convenient, we choose lm() method. However, if we want to look into the details and math properties of linear regression, we can use manual way to get the results.





## Question h: 
do the similar things in (f), compute the result of residual_variance. What the result you observe?

In [75]:
print(true_sigma**2)
print(beta_variance)
print(model.scale)

df = pd.DataFrame({'true': true_sigma**2, 'beta': beta_variance, 'lm': model.scale}, index=["residual_variance"])
print(df)

0.25
0.2462009379126242
0.24657097940247283
                   true      beta        lm
residual_variance  0.25  0.246201  0.246571


To compare the true residual variance σ2 with the estimated residual variance obtained from manual and lm methods, create a matrix with 1 row and 3 columns. For the lm method, get the estimated residual standard deviation from the summary() list object in a list entry called sigma. Convert this to residual variance by squaring it. We can observe that these three constant are nearly the same, which indicates that these three methods get the same results in different ways. There are many ways to do the linear regression analysis. If we want to be more convenient, we choose lm() method. However, if we want to look into the details and math properties of linear regression, we can use manual way to get the results.






## Question i
Based on the results in (f), now, if there is a new student, who has HS_math_GPA = 3.123456, HS_Calculus = 1. and NTU_precalculus=0. What is his/her prediction of NTU_Calculus_I_grade (y) according to you linear regression Truth, manual, and lm, respectively?

In [77]:
new_HS_math_GPA = 3.123456
new_HS_Calculus = 1
new_NTU_precalculus=0


print("truth_score: ",true_beta[0]+new_HS_math_GPA*true_beta[1]+new_HS_Calculus*true_beta[2]+new_NTU_precalculus*true_beta[3])
print("manual_score: ", beta_hat[0]+new_HS_math_GPA*beta_hat[1]+new_HS_Calculus*beta_hat[2]+new_NTU_precalculus*beta_hat[3])
print("lm_score: ", lm_beta[0]+new_HS_math_GPA*lm_beta[1]+new_HS_Calculus*lm_beta[2]+new_NTU_precalculus*lm_beta[3])


truth_score:  2.7864191999999997
manual_score:  2.7885173649046306
lm_score:  2.7885173649046204


We just use the data of each betas, and add all independent variables together. We can see that these three score are nearly the same. Cool!